[QC-326] Aggregators

Barthelemy opened the pull request.

Barthelemy commented:
Still a bit of work to be done on the doxygen.

knopers8 commented on Framework/src/Aggregator.cxx:
```diff
@@ -0,0 +1,120 @@
+void Aggregator::init()
+{
+  try {
+    ILOG(Info, Devel) << "Insantiating the user code for aggregator " << mAggregatorConfig.name
```
```suggestion
    ILOG(Info, Devel) << "Instantiating the user code for aggregator " << mAggregatorConfig.name
```

knopers8 commented on Framework/src/Aggregator.cxx:
```diff
@@ -0,0 +1,120 @@
+  ILOG(Info, Ops) << mAggregatorConfig.name << ": Class " << mAggregatorConfig.className << AliceO2::InfoLogger::InfoLogger::endm;
+  ILOG(Info, Ops) << mAggregatorConfig.name << ": Detector " << mAggregatorConfig.detectorName << AliceO2::InfoLogger::InfoLogger::endm;
+  ILOG(Info, Ops) << mAggregatorConfig.name << ": Policy " << mAggregatorConfig.policyType << AliceO2::InfoLogger::InfoLogger::endm;
+  ILOG(Info, Ops) << mAggregatorConfig.name << ": QualityObjects : " << AliceO2::InfoLogger::InfoLogger::endm;
+  for (const auto& moname : mAggregatorConfig.moNames) {
+    ILOG(Info, Ops) << mAggregatorConfig.name << "   - " << moname << AliceO2::InfoLogger::InfoLogger::endm;
+  }
```
I would also add indication if "all" objects are expected, for the clarity. Just an empty list of QOs might be misleading in such case.

Barthelemy replied:
solved by removing this option, see other comment

knopers8 commented on Framework/src/Aggregator.cxx:
```diff
@@ -0,0 +1,120 @@
+        ILOG(Info, Devel) << "      (no QOs specified or specified as `all`)" << ENDM;
+        mAggregatorConfig.allMOs = true;
+        // fixme: this is a dirty fix. Policies should be refactored, so this check won't be needed.
+        if (mAggregatorConfig.policyType != "OnEachSeparately") {
```
Anyway, does "OnEachSeparately" actually make any sense in case of aggregating? Maybe we should not mention it any examples and throw if it is chosen for an aggregator?

Barthelemy replied:
agreed, I had removed it already from the doc.

knopers8 commented on Framework/src/Aggregator.cxx:
```diff
@@ -0,0 +1,120 @@
+{
+  auto results = mAggregatorInterface->aggregate(qoMap);
+  QualityObjectsType qualityObjects;
+  for (auto const& [name, quality] : results) {
+    qualityObjects.emplace_back(std::make_shared<QualityObject>(
+      quality,
+      mAggregatorConfig.name + "/" + name,
```
```suggestion
  for (auto const& [qualityName, quality] : results) {
    qualityObjects.emplace_back(std::make_shared<QualityObject>(
      quality,
      mAggregatorConfig.name + "/" + qualityName,
```
For clarity.

knopers8 commented on Framework/src/AggregatorRunner.cxx:
```diff
@@ -0,0 +1,227 @@
+
+#include "QualityControl/AggregatorRunner.h"
+
+#include <fairlogger/Logger.h>
```
I don't think we need it here.

knopers8 commented on Framework/src/AggregatorRunner.cxx:
```diff
@@ -0,0 +1,227 @@
+  int i = 0;
+  for (const auto& spec : checkRunnerOutputs) {
+    auto input = DataSpecUtils::matchingInput(spec);
+    input.binding = "checkerOutput" + to_string(i);
```
why not use the `binding` from `spec` to make it simpler and clearer?

Barthelemy replied:
Because it is empty. I had talked to Giulio but I don't remember the exact reason. I have added a comment to explain it.

knopers8 replied:
ok

knopers8 commented on Framework/src/AggregatorRunner.cxx:
```diff
@@ -0,0 +1,227 @@
+{
+  framework::InputRecord& inputs = ctx.inputs();
+  for (auto const& ref : InputRecordWalker(inputs)) { // InputRecordWalker because the output of CheckRunner can be multi-part
+    if (ref.header != nullptr) {
```
InputRecordWalker should walk only on non-empty inputs, I think it is redundant.

knopers8 commented on Framework/src/AggregatorRunner.cxx:
```diff
@@ -0,0 +1,227 @@
+  ILOG(Debug, Trace) << "Aggregate called in AggregatorRunner, QOs in cache: " << mQualityObjects.size() << ENDM;
+
+  QualityObjectsType allQOs;
+  for (auto& aggregator : mAggregatorsMap) {
```
You could use a structured binding here perhaps.

knopers8 reviewed (changes requested):
Thank you. I have a lot of small suggestions and questions, but in total, I think it is very good.
I guess that the next step would be to make some common aggregator like WorstFromAll.

I am still a bit puzzled what should happen an aggregator depends on another aggregator. Is it guaranteed that it will be invoked in the same `run()`? 
I am also wondering about circular dependencies among aggregators - this might be a rare case, but perhaps we should check against that to warn a user. Maybe not in this iteration, but later.

Barthelemy commented on Framework/include/QualityControl/Aggregator.h:
```diff
@@ -0,0 +1,75 @@
+
+  o2::quality_control::core::QualityObjectsType aggregate(core::QualityObjectsMapType& qoMap);
+
+  const std::string& getName() const { return mAggregatorConfig.name; };
```
you are paid for being picky :)

Barthelemy commented:
> Thank you. I have a lot of small suggestions and questions, but in total, I think it is very good.
I guess that the next step would be to make some common aggregator like WorstFromAll.

Great review, thank you !
I am addressing the suggestions step by step. 

Yes, I have create https://alice.its.cern.ch/jira/browse/QC-488. I have assigned it to you for the time being with the idea that it might be beneficial that someone else tries out the aggregators. However, I can take it back if needed.

> I am still a bit puzzled what should happen an aggregator depends on another aggregator. Is it guaranteed that it will be invoked in the same run()?

No, it is not guaranteed. If they are defined in a non optimal order, then you could have to wait for the next call to run(). It means that in this case, the aggregator will have to work with the quality of the last cycle, i.e. that it there will be a kind of delay. We could improve it by creating a tree and making them run in the correct order. I would postpone that for a future improvement but I will add a comment in the documentation. https://alice.its.cern.ch/jira/browse/QC-489

> I am also wondering about circular dependencies among aggregators - this might be a rare case, but perhaps we should check against that to warn a user. Maybe not in this iteration, but later.

I considered circular dependencies and it should not be a disaster for the policy OnAny because the aggregators have access to the quality of the previous cycle in the worst case (we maintain the container). For "OnAll" it could be a problem because it will never trigger. I propose to treat it in QC-489 as well.

knopers8 commented:
Yes, it would be good if I played with it as well, the ticket can stay with me.
Agreed for the other issues as well.

Barthelemy commented on Framework/src/InfrastructureGenerator.cxx:
```diff
@@ -449,11 +452,25 @@ void InfrastructureGenerator::generateCheckRunners(framework::WorkflowSpec& work
+      checkRunnerOutputs.insert(checkRunnerOutputs.end(), spec.outputs.begin(), spec.outputs.end());
     }
   }
+
+  ILOG(Info) << ">> Outputs (" << checkRunnerOutputs.size() << "): ";
+  for (const auto& output : checkRunnerOutputs)
+    ILOG(Info) << DataSpecUtils::describe(output) << " ";
+  ILOG(Info) << ENDM;
+
+  // as we need information about the checkRunners we do it here instead of a separate static method
+  AggregatorRunnerFactory aggregatorRunnerFactory;
+  DataProcessorSpec spec = aggregatorRunnerFactory.create(checkRunnerOutputs, configurationSource);
+  workflow.emplace_back(spec);
```
I agree. I have extracted the static method. However, to avoid recomputing twice the same thing I am returning from generateCheckRunners the list of outputs and passing it to the generateAggregators. I have put a note to look into it again.

knopers8 commented on Framework/include/QualityControl/AggregatorInterface.h:
```diff
@@ -0,0 +1,73 @@
+#include "QualityControl/QualityObject.h"
+#include "QualityControl/Quality.h"
+
+using namespace o2::quality_control::core;
```
sorry, i meant "should not"

knopers8 commented on doc/ModulesDevelopment.md:
```diff
@@ -367,8 +367,7 @@ o2-qc-run-advanced --no-qc | o2-qc --config json://${QUALITYCONTROL_ROOT}/etc/ad
           "type": "Check",
-          "name": "QcCheck",
-          "QOs": ["newQuality", "another"]
+          "name": "QcCheck", "": "If the check produces multiple objects, specify \"MOs\""
```
MOs -> QOs?

knopers8 commented on Framework/src/InfrastructureGenerator.cxx:
```diff
@@ -467,9 +468,13 @@ void InfrastructureGenerator::generateCheckRunners(framework::WorkflowSpec& work
+void InfrastructureGenerator::generateAggregator(WorkflowSpec& workflow, std::string configurationSource, vector<framework::OutputSpec>& checkRunnerOutputs)
+{
+  // TODO consider whether we should recompute checkRunnerOutputs instead of receiving it all baked.
+  DataProcessorSpec spec = AggregatorRunnerFactory::create(checkRunnerOutputs, configurationSource);
```
As an improvement, we probably should use only outputs which are needed by the aggregator

Barthelemy replied:
absolutely: https://alice.its.cern.ch/jira/browse/QC-492

Barthelemy commented on Framework/src/AggregatorRunnerFactory.cxx:
```diff
@@ -0,0 +1,59 @@
+    aggregator.getDeviceName(),
+    aggregator.getInputs(),
+    Outputs{},
+    adaptFromTask<AggregatorRunner>(std::move(aggregator)),
```
indeed, also fixed in CheckRunnerFactory.

Barthelemy commented on Framework/include/QualityControl/CheckConfig.h:
```diff
@@ -22,9 +22,9 @@
 {
 
-/// \brief  Container for the configuration of a Check
+/// \brief  Container for the configuration of a Check or an Aggregator.
```
I don't have a strong opinion on keeping the same structure or not. I hesitated and just went for the quick option. If you feel strongly about it let me know. Meanwhile I will make CheckConfig's names more general.

Barthelemy commented on Framework/advanced-aggregator.json:
```diff
@@ -0,0 +1,155 @@
+          "type": "Check",
+          "name": ["dataSizeCheck"]
+        }, {
+          "type": "Check",  "": "this one is using onEachSeparately and thus it sends under a full path",
```
irrelevant indeed, it was a leftover.

Barthelemy commented on Framework/basic-aggregator.json:
```diff
@@ -0,0 +1,103 @@
+        "policy": "OnAll",
+        "detectorName": "TST",
+        "dataSource": [{
+            "type": "Check",   "": "for a check which produces a single result do it like that",
```
absolutely none :) here my point was just to see that the aggregator is properly launched, connected and produces objects. I see your point, so I will get both checks output.

Barthelemy commented on Framework/basic-aggregator.json:
```diff
@@ -0,0 +1,103 @@
+        "detectorName": "TST",
+        "dataSource": [{
+            "type": "Check",   "": "for a check which produces a single result do it like that",
+            "name": "QcCheck", "": "no need to specify each object because there is only 1"
```
You might have seen some leftovers of a previous version where I had a vector called `names`. I removed it in the end because it was a bit confusing when you wanted to specify objects and you don't know to which Check they belong. This being said, it was far less verbose. I would propose to wait for the request to come.

knopers8 replied:
ok

Barthelemy commented on Framework/include/QualityControl/AggregatorInterface.h:
```diff
@@ -0,0 +1,73 @@
+  ///
+  /// @param qoMap A map of the the QualityObjects to aggregate and their full names.
+  /// @return The new qualities, associated with a name.
+  virtual std::map<std::string, Quality> aggregate(std::map<std::string, std::shared_ptr<const o2::quality_control::core::QualityObject>>& qoMap) = 0;
```
The name is the name associated with the Quality being returned. Note that it is not a `QualityObject` but a `Quality`.

knopers8 replied:
ah, indeed, sorry.

Barthelemy commented on Framework/src/Aggregator.cxx:
```diff
@@ -0,0 +1,120 @@
+      auto sourceName = dataSource.get<std::string>("name");
+      ILOG(Info, Devel) << "   Found a source : " << sourceName << ENDM;
+
+      if (dataSource.count("QOs") == 0 || dataSource.get<std::string>("QOs") == "all") {
```
It does not throw for some reason. I have just tested it again. 
I agree with you though. Thinking more about it, I am not sure we need the "all" at all. Why not have 
1. a list of QOs or
2. no list meaning "all" 
?

This way there is not ambiguity. We have seen already how confusing it is for the users.

Barthelemy replied:
I am going to prepare the commit for that change and we can revert it if need be.

knopers8 replied:
I agree, but then we also should make it the same for checkers.

Barthelemy replied:
ok on it

knopers8 replied:
don't forget to update the doc accordingly

Barthelemy commented:
I think that I fixed most of the things. There are 3 open items that I let you have a look at on Monday.
